File: Leetcode/19. Remove Nth Node From End of List/19_Remove_Node.java
class ListNode {
    int val;
    ListNode next;

    ListNode() {
    }

    ListNode(int val) {
        this.val = val;
    }

    ListNode(int val, ListNode next) {
        this.val = val;
        this.next = next;
    }
}

class Solution {
    public ListNode removeNthFromEnd(ListNode head, int n) {
        ListNode fast = head, slow = head;

        for (int i = 0; i < n; i++)
            fast = fast.next;

        if (fast == null)
            fast = head.next;

        while (fast.next != null) {
            slow = slow.next;
            fast = fast.next;
        }
        slow.next = slow.next.next;
        return head;
    }

    public static void main(String[] args) {
        ListNode head = new ListNode(1);
        ListNode current = head;
        int[] values = { 2, 3, 4, 5 };
        for (int val : values) {
            current.next = new ListNode(val);
            current = current.next;
        }
        Solution solution = new Solution();
        int n = 2;
        ListNode updatedHead = solution.removeNthFromEnd(head, n);
        current = updatedHead;
        while (current != null) {
            System.out.print(current.val + " -> ");
            current = current.next;
        }
        System.out.println("null");
    }
}

File: Leetcode/19. Remove Nth Node From End of List/19_Remove_Node.py
from typing import Optional, List


class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next


class Solution:
    def removeNthFromEnd(self, head: Optional[ListNode], n: int) -> Optional[ListNode]:
        dummy = ListNode(0, head)
        slow = dummy
        fast = head

        while n > 0 and fast:
            fast = fast.next
            n -= 1

        while fast:
            slow = slow.next
            fast = fast.next

        slow.next = slow.next.next
        return dummy.next


if __name__ == "__main__":
    head = ListNode(1)
    current = head
    values = [2, 3, 4, 5]
    for val in values:
        current.next = ListNode(val)
        current = current.next

    solution = Solution()

    n = 2
    updated_head = solution.removeNthFromEnd(head, n)

    current = updated_head
    while current:
        print(current.val, end=" -> ")
        current = current.next
    print("None")
